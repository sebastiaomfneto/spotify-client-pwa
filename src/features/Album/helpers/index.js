/**
 * Returns firsrt artist name or label "various artists"
 * @param {object[]} artists 
 * @returns {string}
 */

export function buildArtist(artists = []) {
  if (artists.length > 1) {
    return 'Various artists';
  }

  return artists[0]?.name;
}

/**
 * Returns availability label by available markets
 *
 * @param {object[]} availableMarkets
 * @returns {string}
 */
export function buildAvailability(availableMarkets = []) {
  return availableMarkets.includes('BR')
    ? 'Available in Brazil'
    : 'Unavailable in Brazil';
}
