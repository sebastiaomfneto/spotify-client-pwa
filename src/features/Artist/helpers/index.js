/**
 * Returns popularity label by followers number
 *
 * @param {number} followers
 * @returns {string}
 */
export function buildPopularity(followers = 0) {
  switch (true) {
    case followers >= 80:
      return 'Hot';
    case followers >= 60 && followers <= 79:
      return 'Cool';
    case followers >= 30 && followers <= 59:
      return 'Regular';
    case followers < 30:
      return 'Underground';
    default:
      return '';
  }
}
