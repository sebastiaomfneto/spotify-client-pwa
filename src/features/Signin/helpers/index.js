const { REACT_APP_SPOTIFY_CLIENT_ID } = process.env;

/**
 * Redirect to spotify to authenticate
 *
 * @returns {void}
 */
export function redirectToAuthenticateOnSpotify() {
  window.location.href = `https://accounts.spotify.com/authorize?client_id=${REACT_APP_SPOTIFY_CLIENT_ID}&response_type=token&redirect_uri=${window.location.origin}/signin`;
}

/**
 * Get token from url hash
 *
 * @param {string} hash
 * @returns {object}
 */
export function getTokenFromHash(hash = '') {
  return hash
    .replace(/^#/, '')
    .split(/&/)
    .map((i) => i.split(/=/))
    .reduce((ag, [k, v]) => ({ ...ag, [k]: v }), {});
}
