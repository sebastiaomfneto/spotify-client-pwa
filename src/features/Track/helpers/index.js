/**
 * Build artists name label
 *
 * @param {object[]} artists
 * @returns {string}
 */
export function buildArtist(artists = []) {
  return artists.map((a) => a.name).join(', ');
}

/**
 * Format duration miliseconds into MM:SS string
 *
 * @param {number} ms
 * @returns string
 */
export function buildDuration(ms = 0) {
  const min = (ms / 1000 / 60) << 0;
  let sec = (ms / 1000) % 60 << 0;

  if (sec < 10) {
    sec = sec.toString() + '0';
  }

  return `${min}:${sec}`;
}
