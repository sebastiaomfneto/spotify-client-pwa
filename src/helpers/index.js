export function buildImageSrc([image] = []) {
  return image?.url;
}
