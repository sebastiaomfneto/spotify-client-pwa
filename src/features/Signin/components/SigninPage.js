import React, { useEffect, useContext } from 'react';
import { useLocation, useHistory } from 'react-router-dom';

import { ReactComponent as SpotifyLogo } from '../../../assets/spotify-logo.svg';
import { PageLayout } from '../../../components';
import { AuthenticationContext, SET_TOKEN } from '../../../contexts';
import { redirectToAuthenticateOnSpotify, getTokenFromHash } from '../helpers';

import './SigninPage.scss';

export function SigninPage() {
  const history = useHistory();

  const { hash } = useLocation();

  const { dispatch } = useContext(AuthenticationContext);

  useEffect(() => {
    if (hash) {
      dispatch({ type: SET_TOKEN, payload: getTokenFromHash(hash) });

      process.nextTick(() => {
        history.push('/artists');
      });
    }
  }, [hash, dispatch, history]);

  return (
    <PageLayout title="Signin">
      <section className="Signin">
        <SpotifyLogo width="150px" height="150px" fill="#1db954" />
        <button onClick={() => redirectToAuthenticateOnSpotify()}>
          Sign In With Spotify
        </button>
      </section>
    </PageLayout>
  );
}
