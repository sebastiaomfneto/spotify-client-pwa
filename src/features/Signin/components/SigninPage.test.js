import React from 'react';
import { render, act, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import * as helpers from '../helpers';
import {
  AuthenticationContext,
  SET_TOKEN,
} from '../../../contexts/AuthenticationContext';
import { SigninPage } from './SigninPage';

jest.mock('../../../services/StorageService');
jest.mock('../helpers');

const withProviders = (
  element,
  initialRoute = '/signin',
  state = '',
  dispatch,
  isAuthenticated = false
) => (
  <MemoryRouter initialEntries={[initialRoute]}>
    <AuthenticationContext.Provider
      value={{ state, dispatch, isAuthenticated }}
    >
      {element}
    </AuthenticationContext.Provider>
  </MemoryRouter>
);

describe('testing SigninPage', () => {
  test('should render login button', async () => {
    jest
      .spyOn(helpers, 'redirectToAuthenticateOnSpotify')
      .mockImplementationOnce(() => {});

    const { baseElement, findByTestId } = render(withProviders(<SigninPage />));

    expect(baseElement).toMatchSnapshot();

    await act(async () => {
      fireEvent.click(await findByTestId('SigninPage-signin-button'));
    });

    expect(helpers.redirectToAuthenticateOnSpotify).toHaveBeenCalled();
  });

  test('should parser location hash and dispatch auth token', () => {
    const token = {
      token_type: 'TEST_TYPE',
      access_token: 'TEST_TOKEN',
    };
    jest.spyOn(helpers, 'getTokenFromHash').mockImplementationOnce(() => token);

    const hash = '#token_type=TEST_TYPE&access_token=TEST_TOKEN';

    const dispatch = jest.fn();

    render(withProviders(<SigninPage />, `/signin${hash}`, '', dispatch));

    expect(dispatch).toHaveBeenLastCalledWith({
      type: SET_TOKEN,
      payload: token,
    });
  });
});
