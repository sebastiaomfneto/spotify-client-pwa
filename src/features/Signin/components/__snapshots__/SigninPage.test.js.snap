// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing SigninPage should render login button 1`] = `
<body>
  <div>
    <div
      class="PageLayout"
    >
      <div>
        <h1>
          Signin
        </h1>
      </div>
      <section
        class="Signin"
      >
        <svg
          fill="#1db954"
          height="150px"
          width="150px"
        >
          spotify-logo.svg
        </svg>
        <button
          data-testid="SigninPage-signin-button"
        >
          Sign In With Spotify
        </button>
      </section>
    </div>
  </div>
</body>
`;
